lib/sdam/server.js

```diff
@@ -39,6 +118,331 @@ class Server extends EventEmitter {
+   * @param {boolean} [options.ordered=true] Execute in order or out of order
+   * @param {object} [options.writeConcern={}] Write concern for the operation
+   * @param {Boolean} [options.serializeFunctions=false] Specify if functions on an object should be serialized.
+   * @param {Boolean} [options.ignoreUndefined=false] Specify if the BSON serializer should ignore undefined fields.
+   * @param {ClientSession} [options.session=null] Session to use for the operation
+   * @param {opResultCallback} callback A callback function
+   */
+  remove(ns, ops, options, callback) {
+    if (typeof options === 'function') {
+      (callback = options), (options = {}), (options = options || {});
+    }
+
+    const error = basicWriteValidations(this, options);
+    if (error) {
+      callback(error, null);
+      return;
+    }
+
+    // Check if we have collation support
+    if (this.description.maxWireVersion < 5 && options.collation) {
+      callback(new MongoError(`server ${this.name} does not support collation`));
+      return;
+    }
+
+    // Setup the docs as an array
+    ops = Array.isArray(ops) ? ops : [ops];
+    // Execute write
+    return this.s.wireProtocolHandler.remove(this.s.pool, ns, this.s.bson, ops, options, callback);
+  }
+}
+
+function basicWriteValidations(server) {
+  if (!server.s.pool) {
+    return new MongoError('server instance is not connected');
+  }
+
+  if (server.s.pool.isDestroyed()) {
+    return new MongoError('server instance pool was destroyed');
+  }
+
+  return null;
+}
+
+function basicReadValidations(server, options) {
+  const error = basicWriteValidations(server, options);
+  if (error) {
+    return error;
+  }
+
+  if (options.readPreference && !(options.readPreference instanceof ReadPreference)) {
+    return new Error('readPreference must be an instance of ReadPreference');
+  }
+}
+
+function saslSupportedMechs(options) {
+  if (!options) {
+    return {};
+  }
+
+  const authArray = options.auth || [];
+  const authMechanism = authArray[0] || options.authMechanism;
+  const authSource = authArray[1] || options.authSource || options.dbName || 'admin';
+  const user = authArray[2] || options.user;
+
+  if (typeof authMechanism === 'string' && authMechanism.toUpperCase() !== 'DEFAULT') {
+    return {};
+  }
+
+  if (!user) {
+    return {};
+  }
+
+  return { saslSupportedMechs: `${authSource}.${user}` };
+}
+
+function extractIsMasterError(err, result) {
+  if (err) return err;
+  if (result && result.result && result.result.ok === 0) {
+    return new MongoError(result.result);
+  }
+}
+
+function executeServerHandshake(server, callback) {
+  // construct an `ismaster` query
+  const compressors =
+    server.s.options.compression && server.s.options.compression.compressors
+      ? server.s.options.compression.compressors
+      : [];
+
+  const queryOptions = { numberToSkip: 0, numberToReturn: -1, checkKeys: false, slaveOk: true };
+  const query = new Query(
```

Comment:
no more than we already had. Much of what you're commenting on was copied right from the old `Server` implementation which had to be modified because of small variations in the grand SDAM refactor. I agree, we are going to have to have a long think about OP_QUERY vs OP_MSG, but I think we should punt until the command refactor, which I think is far more concerned with these types of questions.